### ISSUE TITLE
futile.logger and futures

### ISSUE THREAD

**Issue by adrfantini** [zatonovo/futile.logger#84]
I use `futile.logger` a lot in my parallel scripts, all of which use package `future` to handle multiprocessing.
`flog.*` of course works well in this case when logging to a file, but does not for logging to the terminal: all messages are relayed at the end of the future.

Is there any way this could be accomplished?

**Comment by muxspace** [comment]
Hmm, I'm not sure as I haven't used the `future` package. That said, there is probably a way to force evaluation of a particular call or immediately flush a buffer. It might make more sense to ask the author of that package about how to do that.

**Comment by muxspace** [comment]
@adrfantini Sure that would be great. I'm trying to wrap up my book, so I haven't had much time to look into improvements here. That said, I would like to get a new release to CRAN by end of year, so if there's stuff that needs to be done on this end to support the future package, it would be better to do it sooner than later.

**Comment by adrfantini** [comment]
I'll get around to it in a few days.

**Comment by adrfantini** [comment]
> @adrfantini Can you give me a minimally working example so I can see what you're experiencing? Also, are you writing to the console or to a file?

Sure thing. Writing to the console:
```
library(purrr)
library(furrr)
library(futile.logger)
a = future_map(1:5, function(i) {Sys.sleep(i) ; flog.info('Waited %d seconds', i)} )
```
You will see that the logs will be printed correctly (timings are OK), but they are only flushed at the end of the future.
If, instead, you issue `a = map(1:5, function(i) {Sys.sleep(i) ; flog.info('Waited %d seconds', i)} )`, the logs are of course printed live.

Using `futures`, this even happens if using `plan(sequential)`, so I'm not really sure what's going on.

**Comment by adrfantini** [comment]
Oh, in any case logging to a file works as expected, which is nice.

**Comment by 1beb** [comment]
https://stackoverflow.com/questions/10903787/how-can-i-print-when-using-dopar

**Comment by HenrikBengtsson** [comment]
@adrfantini wrote:

> You will see that the logs will be printed correctly (timings are OK), but they are only flushed at the end of the future.

This is per design of futures; standard output, messages, and warnings are automatically captured and then relayed when the values of the futures are requested.  This is done such that the output of these will be in the same order regardless of parallelization backend, i.e. you will get the exact same output with `plan(sequential)` and `plan(`multisession`) and so on.

As explained in Issue #91, since futile.logger sends log output to the console via `cat()`, that standard output is captured by futures and "only" relayed at the end.  Again, this is by design and on purpose.

My suggestion in Issue #91 is to output via the `message()` framework instead of `cat()`.   The suggestion in Issue #91 will then cause futile.logger to signal log output as `message` conditions.  Now, since the future framework also captures those and "only" relays them at the very end, it will not make a difference to what you're experiencing above ...

However, if you do one more tweak to what is proposed in Issue #91, namely use:

```sh
.outputToConsole <- function(line) {
  msg <- paste(line, collapse = "")
  cond <- simpleMessage(msg, appendLF = FALSE)
  class(cond) <- c("futile.logger.message", "immediateCondition", class(cond))
  message(cond)
}
```
any futile.logger log output to the console will be relayed in a near-live fashion and out of order.  This is because the future framework recognizes the `immediateCondition` class as special and allows them be be signals upstream as soon as possible.

PS. The above is how progression updates signaled by the [**progressr**](https://cran.r-project.org/web/packages/progressr/) package does it, which is why they will also work in future-ized parallel processing.  Even when parallelizing on remote machines.

In summary, with only a minor update to your future.logger code, you can get logging to work in parallel too (when parallelizing using futures).

**Comment by adrfantini** [comment]
Thanks for the clear and complete response. It's very helpful and I consider the ~~case~~ issue closed.

Also thanks for the awesome `progressr` package!

**Comment by muxspace** [comment]
As mentioned in the other thread, I'll see about using `message` instead of `cat`.
